libr/anal/esil_interrupt.c

```diff
@@ -52,34 +73,62 @@ R_API RAnalEsilInterrupt *r_anal_esil_interrupt_new(RAnalEsil *esil, ut32 src_id
 	return intr;
 }
 
+R_API RAnalEsilSyscall *r_anal_esil_syscall_new(RAnalEsil *esil, ut32 src_id, RAnalEsilHandler *sh) {
+	return r_anal_esil_interrupt_new (esil, src_id, sh);
+}
+
 R_API void r_anal_esil_interrupt_free(RAnalEsil *esil, RAnalEsilInterrupt *intr) {
 	if (intr && esil) {
 		_del_interrupt (esil, intr->handler->num);
 	}
 	if (intr) {
 		if (intr->user) {
-			intr->handler->fini (intr->user);	//fini must exist when user is !NULL
+			intr->handler->fini (intr->user); //fini must exist when user is !NULL
 		}
 		r_anal_esil_release_source (esil, intr->src_id);
 	}
 	free (intr);
 }
 
+R_API void r_anal_esil_syscall_free(RAnalEsil *esil, RAnalEsilSyscall *sysc) {
+	if (sysc && esil) {
+		_del_syscall (esil, sysc->handler->num);
+	}
+	if (sysc) {
+		if (sysc->user) {
+			sysc->handler->fini (sysc->user); //fini must exist when user is !NULL
+		}
+		r_anal_esil_release_source (esil, sysc->src_id);
+	}
+	free (sysc);
+}
+
 R_API bool r_anal_esil_set_interrupt(RAnalEsil *esil, RAnalEsilInterrupt *intr) {
 	r_return_val_if_fail (esil && esil->interrupts && intr && intr->handler && intr->handler->cb, false);
 	// check if interrupt is already set
-	RAnalEsilInterrupt *o_intr = _get_interrupt(esil, intr->handler->num);
+	RAnalEsilInterrupt *o_intr = _get_interrupt (esil, intr->handler->num);
 	if (o_intr) {
 		r_anal_esil_interrupt_free (esil, o_intr);
 	}
 	//set the new interrupt
-	return _set_interrupt(esil, intr);
+	return _set_interrupt (esil, intr);
+}
+
+R_API bool r_anal_esil_set_syscall(RAnalEsil *esil, RAnalEsilSyscall *sysc) {
+	r_return_val_if_fail (esil && esil->syscalls && sysc && sysc->handler && sysc->handler->cb, false);
+	// check if interrupt is already set
+	RAnalEsilSyscall *o_sysc = _get_syscall (esil, sysc->handler->num);
+	if (o_sysc) {
+		r_anal_esil_syscall_free (esil, o_sysc);
```

Comment:
its always fine to call free with just NULL. but free functions should take only one parameter. otherwise swap the arguments and put `free` in the middle like this: `r_anal_esil_free_syscall(esil, o_sysc);` or make Syscall hold an instance to ESIL